plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -398,6 +393,157 @@ private S3AsyncClient createS3AsyncClient(AwsCredentialsProvider credentialsProv
+                sliceSize = ByteBuffer.wrap(buffer, bufferPosition, Integer.BYTES).order(LITTLE_ENDIAN).getInt();
+            }
+            bufferPosition += Integer.BYTES;
+
+            verify(0 < sliceSize && sliceSize <= bufferFill - bufferPosition,
+                    "Buffer content should contain at least one complete slice");
+            Slice slice = Slices.wrappedBuffer(buffer, bufferPosition, sliceSize);
+            bufferPosition += sliceSize;
+
+            if (bufferFill - bufferPosition > Integer.BYTES) {
+                sliceSize = ByteBuffer.wrap(buffer, bufferPosition, Integer.BYTES).order(LITTLE_ENDIAN).getInt();
+                if (bufferFill - bufferPosition - Integer.BYTES < sliceSize) {
+                    fillBuffer();
+                }
+            }
+            else {
+                sliceSize = -1;
+                fillBuffer();
+            }
+
+            return slice;
+        }
+
+        @Override
+        public ListenableFuture<Void> isBlocked()
+        {
+            return nonCancellationPropagating(inProgressReadFuture);
+        }
+
+        @Override
+        public long getRetainedSize()
+        {
+            return INSTANCE_SIZE + SizeOf.sizeOf(buffer);
+        }
+
+        @Override
+        public void close()
+        {
+            inProgressReadFuture.cancel(true);
+            currentFile = null;
+        }
+
+        private void fillBuffer()
+        {
+            long fileSize = currentFile.getFileSize();
+            if (fileOffset == fileSize) {
+                verify(bufferPosition == bufferFill,
+                        "bufferPosition (%s) expected to equal bufferFill (%s) when current file has been completely read", bufferPosition, bufferFill);
+                if (!exchangeSourceFiles.hasNext()) {
+                    currentFile = null;
+                    return;
+                }
+                currentFile = exchangeSourceFiles.next();
+                fileSize = currentFile.getFileSize();
+                fileOffset = 0;
+            }
+
+            String key = keyFromUri(currentFile.getFileUri());
+            String bucketName = getBucketName(currentFile.getFileUri());
+            Optional<SecretKey> secretKey = currentFile.getSecretKey();
+
+            byte[] oldBuffer = buffer;
+            this.buffer = new byte[bufferSize];
+            if (oldBuffer != null) {
+                verify(bufferPosition <= bufferFill,
+                        "bufferPosition (%s) expected to be smaller or equal to bufferFill (%s)", bufferPosition, bufferFill);
+                int rest = bufferFill - bufferPosition;
+                arraycopy(oldBuffer, bufferPosition, buffer, 0, rest);
+                bufferPosition = 0;
+                bufferFill = rest;
+            }
+
+            // Make sure S3 read request byte ranges align with part sizes for best performance
+            int readableParts = toIntExact(buffer.length - bufferFill) / partSize;
```

Comment:
`readableParts` was a bit confusing to me as it felt it characterizes file, while it actually characterizes buffer.
Maybe rename to `bufferPartsCapacity` or sth similar?

Reply:
It actually characterizes file. So I feel `readablePart` is quite readable here.

Reply:
It's both. Imagine a file of 12MB, with part size 5MB, then it will be stored in three parts: 5MB-5MB-2MB